table/builder.go

```diff
@@ -424,3 +437,69 @@ func (b *Builder) compressData(data []byte) ([]byte, error) {
 	}
 	return nil, errors.New("Unsupported compression type")
 }
+
+// type use 2 bits, now we have two unused type.
+const (
+	typeUint    = 0
+	typeBytes   = 1
+	typeUnused1 = 2
+	typeUnused2 = 3
+)
+
+type metaEncoder struct {
+	buf []byte
+}
+
+func (e *metaEncoder) appendUint(d uint64, id byte) {
+	e.buf = append(e.buf, id<<2|typeUint)
+	e.buf = append(e.buf, u64ToBytes(d)...)
+}
+
+func (e *metaEncoder) appendBytes(d []byte, id byte) {
+	e.buf = append(e.buf, id<<2|typeBytes)
+	e.buf = append(e.buf, u32ToBytes(uint32(len(d)))...)
+	e.buf = append(e.buf, d...)
+}
+
+type metaDecoder struct {
+	buf    []byte
+	cursor int
+}
+
+func (e *metaDecoder) valid() bool {
+	return e.cursor < len(e.buf)
+}
+
+func (e *metaDecoder) currentIdAndType() (id byte, tp byte) {
+	tag := e.buf[e.cursor]
+	return tag >> 2, tag & 3
+}
+
+func (e *metaDecoder) decodeUint() uint64 {
+	y.Assert(e.buf[e.cursor]&3 == typeUint)
+	e.cursor += 1
+	d := bytesToU64(e.buf[e.cursor:])
+	e.cursor += 8
+	return d
+}
+
+func (e *metaDecoder) decodeBytes() []byte {
+	y.Assert(e.buf[e.cursor]&3 == typeBytes)
+	e.cursor += 1
+	l := int(bytesToU32(e.buf[e.cursor:]))
+	e.cursor += 4
+	d := e.buf[e.cursor : e.cursor+l]
+	e.cursor += l
+	return d
+}
+
+func (e *metaDecoder) skip() {
```

Comment:
Or we can just encode everything in bytes.

Reply:
Yes, the penalty is extra 4 bytes for fixed-length type.